### DISCUSSION TITLE
Help with query performance.

### DISCUSSION THREAD

**Discussion by maxenko** [ravendb/ravendb#14222] [Q&A]
Getting somewhat slow performance with moderately sized dataset.

I have test data of about 4 million records. Standard setup with 3 nodes, 3 cores each - dev license. Running in 3 separate VMs.

Each Record looks like this:

```json
{
    "Id@": "01G350X981ZZW6TKQAT7J9725F",
    "topicId": "01G34YHZEC94ZZGFCWG8XTD1XJ",
    "title": "Readier legal discountability incendiary plight abusiveness.",
    "body": "Bedlinen squirts flowerpot ...",
    "attachments": [],
    "tags": [],
    "timeStamp": "2013-07-15T15:54:33.5433984",
    "@metadata": {
        "@collection": "Posts",
        "Raven-Clr-Type": "Pindap.Types+Post, libpindap"
    }
}
```

There are 4 topics with about 1mil records in each.

The query looks like this (testing with RQL for now):

```rql
from Posts where topicId = "01G34Z6H496EJ2PA0W1B61NNAH" 
and timeStamp < "2022-01-10T10:00:32.2277845"
order by timeStamp desc
limit 50
```

Performance is the same when querying index:
```rql
from index 'Auto/Posts/ByIdAndtimeStampAndtopicId'
where topicId = "01G34VJ1NGZABH72EZRCJP7ER2" and timeStamp < "2022-07-12T01:07:20.5666648"
order by timeStamp desc 
limit 50
```

This returns the result in 500-600ms.

The total timestamp range is between now (2022) and 2012. When querying closer to 2012:
```rql
from Posts where topicId = "01G34Z6H496EJ2PA0W1B61NNAH" 
and timeStamp < "2012-01-10T10:00:32.2277845"
order by timeStamp desc
limit 50
```

Result is returned in 150-250ms.

Why is there a difference closer to lower end of dates?
What can I do to improve the performance overall?

**Comment by ppekrol** [comment]
Hi,

I would suggest (if possible) decreasing the granularity of the timestamp values, this will decrease number of terms significantly. For example you can cut the timestamp just to the day. This can be done during indexing like that:

Map:
```
from p in docs.Posts
select new
{
    TimeStampDay = new DateTime(p.timeStamp.Year, p.timeStamp.Month, p.timeStamp.Day)
}
```

**Comment by maxenko** [comment]
@ppekrol thank you, yeah that is definitely a way to speed it up. I could maybe go down to a second, but not further as it would break sorting in some cases... I see there was a significant optimization done in 5.0:

https://ravendb.net/articles/ravendb-5-0-optimizing-date-range-queries

to convert time strings to ticks for this kind of scenario. The query time grows with dataset unfortunately, so its a bandaid, even when Posts are for a different topicId.

Having switched to abstracted out provider in my app, I am trying RavenDB out as alternative to couchbase, which with a correctly set up index does this query in 25-50ms on the same exact hardware (4 cpus per node, 2 nodes), regardless of dataset size, but it has its own problems also.

I'll keep poking at indexes, but so far every single one I've made myself performs a lot worse than the auto one.

**Comment by ppekrol** [comment]
@maxenko 

We made some optimizations in 5.0, that is true. I'm assuming that you are using 5.2 or newer?

You can consider having two fields in index one for filtering, second one for sorting e.g.

```
from p in docs.Posts
select new
{
    topicId = p.topicId,
    timeStampDay = new DateTime(p.timeStamp.Year, p.timeStamp.Month, p.timeStamp.Day),
    timeStamp = p.timeStamp
}
```

then do a query as follows:

```
from index 'MyIndex' where topicId = "01G34Z6H496EJ2PA0W1B61NNAH" 
and timeStampDay < "2022-01-10T00:00:00.0000000"
order by timeStamp desc
limit 50
```

Have you tried sth like this?

**Comment by maxenko** [comment]
Getting exactly same response times. ~450ms on 2m records. Setting date lower like 2012 reduces time to ~150ms.

**Reply by georgiosd** [reply]
@maxenko I am not sure it will make sure a difference but still worth to be mentioned: the benchmark is not "fair" - the community license is limited to ONE core per node and 2GB of memory. 

If you really want to compare, use a single node with the developer license which can use all four cores and more memory.

Also be sure to read this: https://ayende.com/blog/4784/architecting-in-the-pit-of-doom-the-evils-of-the-repository-abstraction-layer

It really changed my mind about this pattern. 

Do you only have one query per call measured?

**Comment by maxenko** [comment]
@georgiosd I am running dev license cluster with 3 RavenDB servers using 9 processors. It has more resources than CB does (which only has 8 cores between two nodes). I don't have a problem with marginally slower performance due to licensing etc, the time delta with dataset growth is a show stopper unfortunately. There is no difference in same cb/n1ql query if its 1, 2, 6 million records, only index size. Using exact same date format.

Thank you for the article. I understand the problem of confines it may (and will) create, but I really can't be dependent on a single storage solution. CB scares me a bit with their recent changes after going public, this is really why I am looking for an 'in case' switch.

I haven't stress loaded it yet, no. Aside from initial setup confusion, RavenDB is pretty awesome. I really appreciate the straight forward api and all the collections automagic.

**Reply by ppekrol** [reply]
can you add 'include timings()` at the end of that RQL query and return us the timings from the Studio? I would not expect that query to take that long. Please also remember to disable cache in Studio, it is in the Query page under Settings.

**Reply by maxenko** [reply]
@ppekrol Please see screenshot.

test index is:
```
from p in docs.Posts
select new
{
    topicId = p.topicId,
    timeStampDay = new DateTime(p.timeStamp.Year, p.timeStamp.Month, p.timeStamp.Day),
    timeStamp = p.timeStamp
}
```

![Screenshot 2022-05-17 092515](https://user-images.githubusercontent.com/1999105/168861806-37835b35-3f40-47e6-bdef-c014075aed78.png)

**Reply by georgiosd** [reply]
@maxenko now you've made me curious 🤔 

Is the dataset something you could share so I can run some tests?

**Comment by maxenko** [comment]
@ppekrol yes of crourse. I'll rip out the tests code (random data generator) and share the repo. Will have to wait until later today.

**Comment by maxenko** [comment]
@ppekrol please use this for data generation:
https://github.com/maxenko/ravendb_sample_db_generator

set your db name (Program.fs line 40):
```fsharp
store.Database <- "pindap"
```

set number of posts to generate (Program.fs line 52):
```fsharp
|> Seq.take (50_000)
```

Thank you for looking into it!

**Reply by ppekrol** [reply]
I've inserted 5M of Posts.

I'm not sure if this is 1 to 1 with your case, but the 'topicId' does not filter much the set. Which means that `timeStamp` part basically is going over 5M of terms. Is there a possibility to decrease the set on which 'timeStamp' part will filter?

e.g. on my machine the difference in execution time between:

```
from index 'MyIndex' 
where 
topicId = "01G3DKSTBBY2HGKNEQNWW7XPNE" 
and timeStamp < "2022-01-10T00:00:00.0000000"
order by timeStamp desc
include timings() 
limit 50
```

and

```
from index 'MyIndex' 
where 
topicId = "01G3DKSTBBY2HGKNEQNWW7XPNE" 
and timeStamp < "2022-01-10T00:00:00.0000000"
and timeStamp > "2021-01-10T00:00:00.0000000"
order by timeStamp desc
include timings() 
limit 50
```

is ~50%

**Comment by maxenko** [comment]
@ppekrol Thank you for looking at it. Its a solid optimization, I just don't know where the last date for the limit range is.

I guess it comes down to ability to specify ASC/DESC per field in the index itself. Is this possible in RavenDB?

CB equivalent:
```
CREATE INDEX `idx_topicId_dateDESC_t` ON `pindap`(`topicId`,`date` DESC) WHERE ((`t` = "post") and (`topicId` is not null))
```

**Reply by ppekrol** [reply]
It's not possible.

I'm thinking about a different approach.

1. Define an index:

Map:
```
from p in docs.Posts
select new {
    p.topicId,
    timeStampDay = new DateTime(p.timeStamp.Year, p.timeStamp.Month, p.timeStamp.Day),
    timeStamps = new [] { new { p.Id, p.timeStamp }  }
}
```

Reduce:
```
from r in results
group r by new { r.topicId, r.timeStampDay } into g
select new {
    topicId = g.Key.topicId,
    timeStampDay = g.Key.timeStampDay,
    timeStamps = g.SelectMany(x => x.timeStamps).OrderByDescending(x => x.timeStamp)
}
```

2. Query it like follows (this query takes 2-4ms on my machine):

```
from index "MyIndex3" 
where topicId = "01G3DKSTBBY2HGKNEQNWW7XPNE" 
and timeStampDay < "2022-01-10T00:00:00.0000000"
order by timeStampDay desc
limit 50
```

3. On Client-side do the actual filtering with higher granularity for timeStamp and execute

```
string[] ids = // filter here
var posts = session.Load<Post>(ids);
```

This will be making 2 calls to the database, but could decrease you the time significantly.

**Reply by maxenko** [reply]
That is pretty amazing. I have to play around with UI if granularity is to a Day, but it works fine to an Hour. I've changed the index to an Hour, and 6.7m documents (same topicId) are returning within 15-50ms. (this is compared to CB 25-50ms). CPU on this hypervisor is E5-2670V3, so fairly old. Appears to be no real difference as dataset grows, there is a difference when looking below older dates (its faster).

This will work great, thank you @ppekrol Appreciate your help!

**Comment by ayende** [comment]
Let's me also add:
* What *exact* version are you running?
* What is the granularity that you need for those timestamps?

**Reply by maxenko** [reply]
@ayende 5.3.102
I think to an hour will work for all cases. I will have to test.

Thank you!
